app/payment-link-v2/confirm/confirm.controller.js

```diff
@@ -39,35 +102,57 @@ function getPage (req, res, next) {
     productName: product.name
   }
 
-  const summaryElements = []
+  const summaryElements = getSummaryElements(
+    sessionReferenceNumber,
+    sessionAmount,
+    product.reference_label,
+    product.externalId,
+    product.price,
+    res.locals.__p
+  )
 
-  if (sessionReferenceNumber) {
-    summaryElements.push(generateSummaryElement(
-      product.reference_label,
-      sessionReferenceNumber,
-      replaceParamsInPath(paymentLinksV2.reference, product.externalId),
-      HIDDEN_FORM_FIELD_ID_REFERENCE_VALUE
-    ))
-  }
+  data.summaryElements = summaryElements
+  data.confirmPageUrl = replaceParamsInPath(paymentLinksV2.confirm, product.externalId)
 
-  const changeAmountUrl = replaceParamsInPath(paymentLinksV2.amount, product.externalId)
-  const totalToPayText = res.locals.__p('paymentLinksV2.confirm.totalToPay')
+  return response(req, res, 'confirm/confirm', data)
+}
 
-  const getAmountToDisplay = getRightAmountToDisplayAsGbp(sessionAmount, product.price)
+async function postPage (req, res, next) {
+  const product = req.product
 
-  summaryElements.push(generateSummaryElement(
-    totalToPayText,
-    getAmountToDisplay,
-    changeAmountUrl,
-    HIDDEN_FORM_FIELD_ID_AMOUNT
-  ))
+  const sessionReferenceNumber = getSessionVariable(req, 'referenceNumber')
+  const sessionAmount = getSessionVariable(req, 'amount')
+
+  const data = {
+    productExternalId: product.externalId,
+    productName: product.name
+  }
+
+  const summaryElements = getSummaryElements(
+    sessionReferenceNumber,
+    sessionAmount,
+    product.reference_label,
+    product.externalId,
+    product.price,
+    res.locals.__p
+  )
```

Comment:
As we're storing these in hidden fields on the page, would it make sense to get them out of there for displaying errors back to the user?
The answer to this isn't necessarily yes. But we're storing them there to prevent session interference if the user starts another payment in another tab. If we're displaying back errors, it might be good to make sure 100% we're displaying back the values they've just seen and checked before attempting the reCAPTCHA.

What are your thoughts?

Reply:
Agree - have updated it to use the values from the `hidden` form fields on the POST method.
Good idea - thanks